### DISCUSSION TITLE
Issue with I2C communication

### DISCUSSION THREAD

**Discussion by dobodu** [micropython/micropython#11756] [RP2040 / Pico] [answered]
Hi,

I've be rewriting the BNO085 circuitpython library in micropython in order to fullfill a project I made with a Pico Pi W.

I'm struggling with Buffer and I2c communication. I believe I miss something when adapting circuitpython I2C behaviour to micropython i2c behaviour.

Here's the commented code : 

```
def _read_packet(self):
self._dbg("BNO08X_I2C READ PACKET :")
self._dbg("")

#Read a packet packet = header + packet data
#Header is 4 bytes long (2 bytes size, 1 byte for channel number and 1 byte for sequence number)
#Buffer is declared in classe : self._data_buffer = bytearray(512)
#I2C adress is declared in class : self.bno_address = const(0x4A)
#Begin with reading a header ==> Expecting a header (4 bytes)
#Should be readfrom_into(self.bno_address, self._data_buffer[0:4])
#but does not give the same values compared to readfrom_into(self.bno_address, self._data_buffer)

self.bus_device_obj.readfrom_into(self.bno_address, self._data_buffer**[0:4]**)
        
self._dbg("\tself._data_buffer[0:4] contains : ", self._data_buffer[0:4])
self._dbg("")
        
header = Packet.header_from_buffer(self._data_buffer[0:4])  #tested and working correctly
packet_byte_count = header.packet_byte_count   		# Working
channel_number = header.channel_number			# Working
sequence_number = header.sequence_number		# Working
data_length = header.data_length 					# Working, data_lenth equals packet_bytes_count minus 4 (header length)

self._sequence_number[channel_number] = sequence_number		# Seq num stored sorted by channel num.

if packet_byte_count == 0:
    self._dbg("\tSKIPPING NO PACKETS AVAILABLE IN i2c._read_packet")
    self._dbg("")
    raise PacketError("No packet available")

self._dbg("\tChannel", channel_number, "has", data_length, "bytes available to read")

#Consequence, normaly next thing is to read the packet data but of course when reading
#the full size buffer before, the reading is incorrect
#Buffer must be after this sequence
#   0  1  2  3  4  5  6  7  8  9 ...
#   **HEADER**  **PACKET DATA******
#
self.bus_device_obj.readfrom_into(self.bno_address, self._data_buffer[4:packet_byte_count])

#Then process the packet : Tested and working

new_packet = Packet(self._data_buffer[0:packet_byte_count])#,True)
if self._debug:
     print(new_packet)

self._update_sequence_number(new_packet)
return new_packet
```

And here's the output

With self.bus_device_obj.readfrom_into(self.bno_address, self._data_buffer)

> DBG::		 BNO08X_I2C READ PACKET :
> DBG::		 
> DBG::		 	self._data_buffer[0:4] contains :  bytearray(b'\x14\x01\x00\x00')
> DBG::		 
> DBG::		 	Channel 0 has 272 bytes available to read
> DBG::
> DBG::			 HEADER:
> DBG::				Data Len: 272
> DBG::				Channel: SHTP_COMMAND (0)
> DBG::				Sequence number: 0
> DBG::			 DATA:
> DBG::				[0x04] 0x00 0x01 0x04 0x00 
> DBG::				[0x08] 0x00 0x00 0x00 0x80 
> DBG::				[0x0C] 0x06 0x31 0x2E 0x30 
> DBG::				[0x10] 0x2E 0x30 0x00 0x02 
> DBG::				[0x14] 0x02 0x00 0x01 0x03 
> DBG::				[0x18] 0x02 0xFF 0x7F 0x04 
> DBG::				[0x1C] 0x02 0x00 0x01 0x05 
> DBG::				[0x20] 0x02 0xFF 0x7F 0x08 

With self.bus_device_obj.readfrom_into(self.bno_address, self._data_buffer[0:4])

> DBG::		 BNO08X_I2C READ PACKET :
> DBG::		 
> DBG::		 	self._data_buffer[0:4] contains :  bytearray(b'\x05\x00\x01\x01')
> DBG::		 
> DBG::		 	Channel 1 has 1 bytes available to read
> DBG::
> DBG::			 HEADER:
> DBG::				Data Len: 1
> DBG::				Channel: EXE (1)
> DBG::				Sequence number: 1
> DBG::			 DATA:
> DBG::				[0x04] 0x01`

It seems that when reading **self._data_buffer[0:4]** the buffer is not updated with the readings I have with  **self._data_buffer**
When using **self._data_buffer[0:4]** the buffer contains the data I used just before (sending the packet bytearray(b'\x05\x00\x01\x01')

I think using "partial" buffer might be a wrong way, everything acts as if the global buffer was not updated.

Any suggestion ?

**Comment by robert-hh** [comment, marked as answer]
self._data_buffer[0:4] creates a new bytearray object, which is different form buffer. So it you read into that one, buffer is not changed. It should work using memoryview, which creates an object which effectively is part of buffer. Not tested:

mv = memoryview(buffer)
then use readinto with mv[0:4] as argument.

**Comment by dobodu** [comment]
Thank you very much Robert, that did the job !

I was thinking of a local vs global issue, but I miss practice => I didn't realized that I was working on a new bytearray object !

My BNO085X library inspired by CircuitPython Orginal Library is fully working right now !
